<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Specific Analyses for MFM package • MFMextra </title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Specific Analyses for MFM package">
<meta property="og:description" content="Several specific simulation-based analyses are provided here.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">MFMextra </a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/MFMextra.html">MFMextra: examples of generating simulated data to evaluate MFM</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jennasimit/MFMextra">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    

    
    
<div id="mfmextra" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#mfmextra" class="anchor"></a>MFMextra</h1></div>
<p>This R package provides the simulation functions used to assess the joint fine-mapping methods of <a href="https://jennasimit.github.io/MFM/">MFM</a>.</p>
<p>Website available at: <a href="https://jennasimit.github.io/MFMextra/" class="uri">https://jennasimit.github.io/MFMextra/</a></p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"devtools"</span>) <span class="co"># if you don't already have the package</span></a>
<a class="sourceLine" id="cb1-2" title="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(devtools)</a>
<a class="sourceLine" id="cb1-3" title="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/devtools/topics/reexports">install_github</a></span>(<span class="st">"jennasimit/MFMextra"</span>)</a></code></pre></div>
</div>
<div id="simulation-of-case-control-data-for-one-disease" class="section level2">
<h2 class="hasAnchor">
<a href="#simulation-of-case-control-data-for-one-disease" class="anchor"></a>Simulation of case-control data for one disease</h2>
<p>Below is an example of how to use <a href="http://mathgen.stats.ox.ac.uk/genetics_software/hapgen/hapgen2.html">hapgen2</a> to simulate N0 cases and N1 controls in a gene (IL2RA) on chromosome 10, where we set two causal variants. The positions of the causal variants are $c1 and $c2, which have $ORhet1 and $ORhet2 as the respective odds ratios relating the odds of disease in heterozygote carriers of the non-reference allele (this is specified by the ‘1’ following each of $c1 and $c2; ‘0’ indicates with respect to the reference allele) compared to the homozygote reference allele. We assume a multiplicative model so that OR for homozygote disease risk is OR^2. For each simulation the causal variants were randomly selected from a certain SNP group.</p>
<p>This requires map, legend and hap files, which could be obtained from a reference panel (e.g. <a href="ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/">CEU of 1000 Genomes</a> vcf file by using <a href="https://vcftools.github.io/index.html">vcftools</a> with the –IMPUTE option. To output only a subset of SNPs, we specify the SNP positions, line-by-line, in the file keep-snps.txt.</p>
<pre><code>ORhom1=$(bc -l &lt;&lt;&lt; "$ORhet1 ^2")
ORhom2=$(bc -l &lt;&lt;&lt; "$ORhet2 ^2")

./hapgen2 \
-m ./genetic_map_chr10_combined_b37.txt #map file \
-l ./IL2RA.impute.legend # legend file \
-h ./IL2RA.impute.hap # hap file \
-n $N0 $N1 # N0 controls, N1 cases \
-dl $c1 1 $ORhet1 $ORhom1 $c2 1 $ORhet2 $ORhom2 \
-no_haps_output -no_gens_output \
-t ./keep-snps.txt  # output for subset of SNPs that are listed in this file \
-Ne 11418 # effective population size; 11418 recommended for CEU \
-o ./CCdata  # prefix of output files</code></pre>
</div>
<div id="simulation-of-null-data-that-forms-the-basis-for-simulating-a-specific-region-for-two-diseases-with-shared-controls" class="section level2">
<h2 class="hasAnchor">
<a href="#simulation-of-null-data-that-forms-the-basis-for-simulating-a-specific-region-for-two-diseases-with-shared-controls" class="anchor"></a>Simulation of null data that forms the basis for simulating a specific region for two diseases with shared controls</h2>
<p>Below is an example of how to use <a href="http://mathgen.stats.ox.ac.uk/genetics_software/hapgen/hapgen2.html">hapgen2</a> to simulate 100,000 individuals with no associated SNPs in a gene (IL2RA) on chromosome 10. This requires map, legend and hap files, which could be obtained from a reference panel (e.g. <a href="ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/">CEU of 1000 Genomes</a> vcf file by using <a href="https://vcftools.github.io/index.html">vcftools</a> with the –IMPUTE option. By not specifying the causal variants, the default is to simulate null data. To output only a subset of SNPs, we specify the SNP positions, line-by-line, in the file keep-snps.txt.</p>
<pre><code>./hapgen2 \
-m ./genetic_map_chr10_combined_b37.txt #map file \
-l ./IL2RA.impute.legend # legend file \
-h ./IL2RA.impute.hap # hap file \
-n 100000 0 # 100,000 controls, 0 cases \
-no_haps_output -no_gens_output \
-t ./keep-snps.txt  # output for subset of SNPs that are listed in this file \
-Ne 11418 # effective population size; 11418 recommended for CEU \
-o ./null_100k  # prefix of output files</code></pre>
<p>Run the following to generate the MFMextra website:</p>
<pre><code>Rscript -e "pkgdown::build_site()"</code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/jennasimit/MFMextra">https://​github.com/​jennasimit/​MFMextra</a>
</li>
<li>Report a bug at <br><a href="https://github.com/jennasimit/MFMextra/issues">https://​github.com/​jennasimit/​MFMextra/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>GPL</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Jenn Asimit <br><small class="roles"> Author, maintainer </small>  </li>
<li>Chris Wallace <br><small class="roles"> Author </small>  </li>
</ul>
</div>

      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Jenn Asimit, Chris Wallace.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
