<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>MFMextra: examples of generating simulated data to evaluate MFM • MFMextra </title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="MFMextra: examples of generating simulated data to evaluate MFM">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MFMextra </a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/MFMextra.html">MFMextra: examples of generating simulated data to evaluate MFM</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jennasimit/MFMextra">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>MFMextra: examples of generating simulated data to evaluate MFM</h1>
                        <h4 class="author">Jenn Asimit</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:ja628@cam.ac.uk" class="email">ja628@cam.ac.uk</a>
      
                              <h4 class="author">Chris Wallace</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:cew54@cam.ac.uk" class="email">cew54@cam.ac.uk</a>
      
                  
            <h4 class="date">2019-02-15</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/jennasimit/MFMextra/blob/master/vignettes/MFMextra.Rmd"><code>vignettes/MFMextra.Rmd</code></a></small>
      <div class="hidden name"><code>MFMextra.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p><a href="https://github.com/jennasimit/MFM">MFM</a> is a package to simultaneously fine-map (select most likely set of causal variants) multiple related diseases with the same set of controls and share information between them. It relies on output from the package <a href="https://github.com/chr1swallace/GUESSFM">GUESSFM</a>, which fine-maps a single disease via stochastic search in a Bayesian framework using <a href="http://www.bgx.org.uk/software/guess.html">GUESS</a>.</p>
<p>This vignette illustrates how to simulate data for two diseases with shared controls using functions in MFMextra.</p>
</div>
<div id="simulation-example" class="section level2">
<h2 class="hasAnchor">
<a href="#simulation-example" class="anchor"></a>Simulation Example</h2>
<p>Genotype and phenotype data of two diseases with shared controls may be simulated using <a href="https://github.com/jennasimit/MFMextra">MFMextra</a>, together with <a href="http://mathgen.stats.ox.ac.uk/genetics_software/hapgen/hapgen2.html">hapgen2</a> and reference panel data (e.g.  <a href="ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/">CEU of 1000 Genomes</a>.</p>
<p>First, we need to simulate some null data from which we will sample to generate the two sets of cases with shared controls. The entire genetic region of interest is simulated to maintain the linkage disequilibrium (LD) structure. In running simulations, there are typically some models of interest that are selected based on analysis of the data. For example, in a <a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1005272">previous fine-mapping</a> of IL2RA in a large international sample, several SNP groups were identified as having the majority of the association signals with the autoimmune diseases multiple sclerosis (MS) and type 1 diabetes (T1D). These groups, together with previously identified lead SNPs for other autoimmune diseases (autoimmune thyroid disease (ATD; rs706799), alopecia areata (AA; rs3118470), rheumatoid arthritis (RA; rs10795791), and ulcerative colitis (UC; rs4147359)) will compose models that contribute to the non-negligible posterior probabilities. Therefore, for computational efficiency, we extract these SNPs from the generated data and focus on these in the fine-mapping simulation analysis.</p>
<p>Below is an example of hapgen2 code to simulate the IL2RA region based on the <a href="ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/">CEU of 1000 Genomes</a> reference panel, where keep-snps.txt would be the list of snp positions that are to be retained.</p>
<pre><code>    ./hapgen2
    -m ./genetic_map_chr10_combined_b37.txt \
    -l ./IL2RA.impute.legend \
    -h ./IL2RA.impute.hap \
    -n 100000 0 # 100,000 \
    -no_haps_output -no_gens_output \
    -t ./keep-snps.txt \
    -Ne 11418 \
    -o ./null_100k \</code></pre>
<p>Here is an example where diseases 1 and 2 have two causal variants of which one is shared: both have causal variant rs61839660 in group A; disease 1 has additional causal variant rs56382813 in group D; disease 2 has additional causal variant rs11594656 in group C. The file null_100k.controls.gen is output from the above hapgen code and needed below. An example of data output is provided with this vignette, where for efficiency the convert.fn step has been run.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1">    <span class="co">#g0 &lt;- read.table(null_100k.controls.gen,header=FALSE,as.is=TRUE)</span></a>
<a class="sourceLine" id="cb2-2" title="2">    <span class="co">#Nn &lt;- (dim(g0)[2]-5)/3</span></a>
<a class="sourceLine" id="cb2-3" title="3">    <span class="co">#snpG &lt;- convert.fn(g0) # convert to a genotype matrix (snp rows, indiv cols)</span></a>
<a class="sourceLine" id="cb2-4" title="4"></a>
<a class="sourceLine" id="cb2-5" title="5"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(MFMextra)    </a>
<a class="sourceLine" id="cb2-6" title="6"><span class="co">#&gt; Warning: replacing previous import 'data.table::melt' by 'reshape::melt'</span></a>
<a class="sourceLine" id="cb2-7" title="7"><span class="co">#&gt; when loading 'GUESSFM'</span></a>
<a class="sourceLine" id="cb2-8" title="8"><span class="co"># snpG included with MFMextra    </span></a>
<a class="sourceLine" id="cb2-9" title="9">    dis &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"AD"</span>,<span class="st">"AC"</span>)</a>
<a class="sourceLine" id="cb2-10" title="10">    c12 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">grep</a></span>(<span class="st">"rs61839660"</span>,<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(snpG)) <span class="co"># A SNP for both diseases</span></a>
<a class="sourceLine" id="cb2-11" title="11">    c1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">grep</a></span>(<span class="st">"rs56382813"</span>,<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(snpG))  <span class="co"># D SNP for disease 1</span></a>
<a class="sourceLine" id="cb2-12" title="12">    c2 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">grep</a></span>(<span class="st">"rs11594656"</span>,<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(snpG))  <span class="co"># C SNP for disease 2</span></a>
<a class="sourceLine" id="cb2-13" title="13">    </a>
<a class="sourceLine" id="cb2-14" title="14">    causals1.ind &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(c12,c1)</a>
<a class="sourceLine" id="cb2-15" title="15">    causals2.ind &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(c12,c2)</a>
<a class="sourceLine" id="cb2-16" title="16">    prev &lt;-<span class="st"> </span><span class="fl">0.1</span> <span class="co"># prevalence for purpose of method evaluation</span></a>
<a class="sourceLine" id="cb2-17" title="17">    </a>
<a class="sourceLine" id="cb2-18" title="18">    N0 &lt;-<span class="st"> </span><span class="dv">3000</span> <span class="co"># disease 1 size</span></a>
<a class="sourceLine" id="cb2-19" title="19">    N1 &lt;-<span class="st"> </span><span class="dv">3000</span> <span class="co"># disease 2 size</span></a>
<a class="sourceLine" id="cb2-20" title="20">    N2 &lt;-<span class="st"> </span><span class="dv">3000</span> <span class="co"># controls size</span></a>
<a class="sourceLine" id="cb2-21" title="21">    ND=<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/vector">vector</a></span>(<span class="st">"list"</span>,<span class="dv">2</span>) <span class="co"># vector of sizes for cases</span></a>
<a class="sourceLine" id="cb2-22" title="22">    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(ND) &lt;-<span class="st"> </span>dis</a>
<a class="sourceLine" id="cb2-23" title="23">    ND[[<span class="dv">1</span>]]&lt;-N1</a>
<a class="sourceLine" id="cb2-24" title="24">    ND[[<span class="dv">2</span>]]&lt;-N2</a>
<a class="sourceLine" id="cb2-25" title="25">    </a>
<a class="sourceLine" id="cb2-26" title="26">    </a>
<a class="sourceLine" id="cb2-27" title="27">    OR1a &lt;-<span class="st"> </span><span class="fl">1.4</span> <span class="co"># OR for A, disease 1</span></a>
<a class="sourceLine" id="cb2-28" title="28">    OR2a &lt;-<span class="st"> </span><span class="fl">1.25</span>  <span class="co"># OR for D, disease 1</span></a>
<a class="sourceLine" id="cb2-29" title="29">    OR1b &lt;-<span class="st"> </span><span class="fl">1.4</span> <span class="co"># OR for A, disease 2</span></a>
<a class="sourceLine" id="cb2-30" title="30">    OR2b &lt;-<span class="st"> </span><span class="fl">1.25</span>  <span class="co"># OR for C, disease 2</span></a>
<a class="sourceLine" id="cb2-31" title="31">    </a>
<a class="sourceLine" id="cb2-32" title="32">    sim &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/MFMextra/topics/phen.gen.fn">phen.gen.fn</a></span>(<span class="dt">beta1=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log</a></span>(prev),<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log</a></span>(OR1a),<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log</a></span>(OR2a)),<span class="dt">beta2=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log</a></span>(prev),<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log</a></span>(OR1b),<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log</a></span>(OR2b)),<span class="dt">snpG=</span>snpG,<span class="dt">N0=</span>N0,<span class="dt">N1=</span>N1,<span class="dt">N2=</span>N2,causals1.ind,causals2.ind)</a>
<a class="sourceLine" id="cb2-33" title="33">    Gm &lt;-<span class="st"> </span><span class="kw">new</span>(<span class="st">"SnpMatrix"</span>,(sim<span class="op">$</span>G<span class="op">+</span><span class="dv">1</span>)) <span class="co"># snp cols, indivs rows # convert to SnpMatrix format, needed for GUESSFM</span></a>
<a class="sourceLine" id="cb2-34" title="34"><span class="co">#&gt; coercing object of mode  numeric  to SnpMatrix</span></a>
<a class="sourceLine" id="cb2-35" title="35">    Gm</a>
<a class="sourceLine" id="cb2-36" title="36"><span class="co">#&gt; A SnpMatrix with  9000 rows and  26 columns</span></a>
<a class="sourceLine" id="cb2-37" title="37"><span class="co">#&gt; Row names:  control.1 ... case2.3000 </span></a>
<a class="sourceLine" id="cb2-38" title="38"><span class="co">#&gt; Col names:  rs12722563 ... rs41295159</span></a></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#simulation-example">Simulation Example</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Jenn Asimit, Chris Wallace.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
